src/measurement.lisp

```diff
@@ -4,27 +4,28 @@
 
 (in-package #:qvm)
 
-(defun force-measurement (measured-value qubit qvm excited-probability)
-  "Force the quantum system QVM to have the qubit QUBIT collapse/measure to MEASURED-VALUE. Modify the amplitudes of all other qubits accordingly.
+(defgeneric force-measurement (measured-value qubit state excited-probability)
+  (:documentation "Manipulate the STATE as to force the QUBIT in STATE to collapse to the MEASURED-VALUE. EXCTIED-PROBABILITY is the probability that the specified QUBIT measures to |1>, regardless of the MEASURED-VALUE it is being forced to."))
 
-EXCITED-PROBABILITY should be the probability that QUBIT measured to |1>, regardless of what it's being forced as.
+(defmethod force-measurement (measured-value qubit (state pure-state) excited-probability)
+  "Force the quantum system of the PURE-STATE STATE to have the qubit QUBIT collapse/measure to MEASURED-VALUE. Modify the amplitudes of all other qubits accordingly. EXCITED-PROBABILITY should be the probability that QUBIT measured to |1>, regardless of what it's being forced as.
 "
   (declare #.*optimize-dangerously-fast*
            (type bit measured-value)
            (type (flonum 0) excited-probability)
            (type nat-tuple-element qubit))
-  (let* ((wavefunction (amplitudes qvm))
+  (let* ((wavefunction (state-elements state))
          (annihilated-state (- 1 measured-value))
          (inv-norm (if (zerop annihilated-state)
                        (/ (sqrt excited-probability))
                        (/ (sqrt (- (flonum 1) excited-probability))))))
-    (declare (type quantum-state wavefunction)
```

Comment:
why is this invalid? It would be Not Good if the wavefunction wasn't a `quantum state` anymore.